rules/firestore.json

```diff
@@ -45,40 +51,132 @@ service cloud.firestore {
+                   request.resource.data.verified == false ||
+                   (userIsVerified() && resource.data.username != null);
         }
-        function userVerified() {
-            return isAuthentified() &&
+        match /parties/{id} {
+            // Anybody can create a part
+            allow create: if userIsVerified();
+            // The owner of a part can delete it
+            allow delete: if userIsVerified() &&
+                             creatorOf(id).id == request.auth.uid;
+            // Anyone can delete a part if the owner is not observing it or has timed out
+            allow delete: if userIsVerified() &&
+                             notObservingOrTimedOut(creatorOf(id), id);
+            // A number of fields are readonly:
+            // - typeGame
+            // - playerZero/playerOne once the part has started
+            // - beginning once the part has started
+            // the part is considered to be started as soon as beginning is set
+            // For any other fields, both players can modify them
+            allow update: if userIsVerified() &&
+                             preservesField('typeGame') &&
+                             (preservesField('playerZero') || resource.data.beginning == null) &&
+                             (preservesField('playerOne') || resource.data.beginning == null) &&
+                             (preservesField('beginning') || resource.data.beginning == null) &&
+                             (resource.data.playerZero == usernameOf(request.auth.uid) ||
+                              resource.data.playerOne == usernameOf(request.auth.uid));
+            // All verified user can read parts (needed for the lobby)
+            allow read: if userIsVerified();
+        }
+        function notObservingOrTimedOut(creator, partId) {
+            let creatorUser =  get(/databases/$(database)/documents/joueurs/$(creator.id));
+            let notObserving = creatorUser.data.observedPart != partId;
+            let timedOut = duration.value(request.time.toMillis(), 'ms') > duration.value(creatorUser.data.last_changed.seconds, 's') + duration.value(10, 's');
+            return notObserving || timedOut;
+        }
+        match /joiners/{partId} {
+            // Anybody can create a joiner as long as they are themselves creator and if the corresponding part exists
+            allow create: if userIsVerified() &&
+                             userIsCreator(request.resource.data) &&
+                             exists(/databases/$(database)/documents/parties/$(partId));
+            // The creator can delete a joiner
+            allow delete: if userIsVerified() && userIsCreator(resource.data);
+            // Anybody else can delete a joiner only if the creator is offline
+            allow delete: if userIsVerified() && notObservingOrTimedOut(creatorOf(partId), partId);
+            // The creator can change all fields except changing partStatus to STARTED (3)
```

Comment:
le commentaire est un peu différent du code, c'est pas "the creator can change all fields until the part is declared start and the field partStatus can only be put to STARTED(3) by the chosen oppoent" ?

Reply:
tiens oui moyen de rafiner cela: créateur peut changer tout (SAUF mettre partStatus à 3) tant que partStatus < 3 (fait)

la partie sur les non-créateurs c'est dans la suite du code